File: IrrgartenP3/irrgartenRuby/UI/textUI.rb
require 'io/console'
require_relative '../irrgarten/directions'

module UI
  class TextUI

    #https://gist.github.com/acook/4190379
    def read_char
      STDIN.echo = false
      STDIN.raw!

      input = STDIN.getc.chr
      if input == "\e"
        input << STDIN.read_nonblock(3) rescue nil
        input << STDIN.read_nonblock(2) rescue nil
      end
    ensure
      STDIN.echo = true
      STDIN.cooked!

      return input
    end

    def next_move
      print "Where? "
      got_input = false
      while (!got_input)
        c = read_char
        case c
        when "\e[A"
          puts "UP ARROW"
          output = Irrgarten::Directions::UP
          got_input = true
        when "\e[B"
          puts "DOWN ARROW"
          output = Irrgarten::Directions::DOWN
          got_input = true
        when "\e[C"
          puts "RIGHT ARROW"
          output = Irrgarten::Directions::RIGHT
          got_input = true
        when "\e[D"
          puts "LEFT ARROW"
          output = Irrgarten::Directions::LEFT
          got_input = true
        when "\u0003"
          puts "CONTROL-C"
          got_input = true
          exit(1)
        else
          #Error
        end
      end
      output
    end

    # Método para mostrar el estado del juego
    def show_game(game_state)
      # Mostrar el laberinto
      puts "===== Labyrinth ====="
      puts game_state[:labyrinth]

      # Mostrar la lista de jugadores
      puts "\n===== Players ====="
      puts game_state[:players]

      # Mostrar la lista de monstruos
      puts "\n===== Monsters ====="
      puts game_state[:monsters]

      # Mostrar el jugador actual
      puts "\nCurrent Player: Player ##{game_state[:current_player_idx]}"

      # Mostrar el estado del juego (si hay un ganador)
      if game_state[:finished]
        puts "\nThere is a winner!"
      else
        puts "\nThe game is still ongoing."
      end

      # Mostrar el registro de eventos del juego
      puts "\n===== Game Log ====="
      puts game_state[:log]
    end
  end
end

File: IrrgartenP5/IrrgartenJava/src/irrgarten/UI/TextUI.java

package irrgarten.UI;

import irrgarten.*;
import irrgarten.GameState;
import java.util.Scanner;


public class TextUI implements UI{
    
    private static Scanner in = new Scanner(System.in);
    
    private char readChar() {
        String s = in.nextLine();     
        return s.charAt(0);
    }
    

    @Override
    public Directions nextMove() {
        System.out.print("Where? ");
        
        Directions direction = Directions.DOWN;
        boolean gotInput = false;
        
        while (!gotInput) {
            char c = readChar();
            switch(c) {
                case 'w':
                    System.out.print(" UP\n");
                    direction = Directions.UP;
                    gotInput = true;
                    break;
                case 's':
                    System.out.print(" DOWN\n");
                    direction = Directions.DOWN;
                    gotInput = true;
                    break;
                case 'd':
                    System.out.print("RIGHT\n");
                    direction = Directions.RIGHT;
                    gotInput = true;
                    break;
                case 'a':
                    System.out.print(" LEFT\n");
                    direction = Directions.LEFT;
                    gotInput = true;    
                    break;
            }
        }    
        return direction;
    }
    
    @Override
    public void showGame(GameState gameState) {
        // Mostrar el laberinto
        System.out.println("===== Labyrinth =====");
        System.out.println(gameState.getLabyrinth());

        // Mostrar información de los jugadores
        System.out.println("\n===== Players =====");
        System.out.println(gameState.getPlayers());

        // Mostrar información de los monstruos
        System.out.println("\n===== Monsters =====");
        System.out.println(gameState.getMonsters());

        // Mostrar el jugador actual
        System.out.println("\nCurrent Player: Player #" + gameState.getCurrentPlayer());

        // Mostrar si hay un ganador
        if (gameState.isWinner()) {
            System.out.println("\nThere is a winner!");
        } else {
            System.out.println("\nThe game is still ongoing.");
        }

        // Mostrar el registro del juego
        System.out.println("\n===== Game Log =====");
        System.out.println(gameState.getLog());
    }

    @Override
    public void disableInteraction() {
        throw new UnsupportedOperationException("FUncion para gui"); // Generated from nbfs://nbhost/SystemFileSystem/Templates/Classes/Code/GeneratedMethodBody
    }



}
